junctiontree/computation.py

```diff
@@ -71,22 +71,25 @@ def get_message(sepset_ix, tree, beliefs, clique_vars):
             for var in clique_vars[ss_ix]
         ]
 
-        neighbor_vars = list(set(neighbor_vars))
+        neighbor_vars = np.unique(neighbor_vars)
 
         # multiply neighbor messages
-        messages = messages if len(messages) else [1]
 
-        msg_prod = dl.einsum(
-                                *messages,
-                                neighbor_vars
+        msg_prod = 1 if len(messages) == 0 else dl.einsum(
+                                                    *messages,
+                                                    neighbor_vars
```

Comment:
Is this special case `1` related only to the normal sum-product? If some other distributive law was used, would the value be different then? If so, should this if-else be inside the specific "sum-product distributive law" so it wouldn't affect all distributive laws? Or, alternatively, should each distributive law define an identity element ("empty value") that is used in this kind of cases? My gut feeling is that this latter option would make perfect sense.

Reply:
Also, now that this if-else is only here, it won't fix all other places where `dl.einsum` is used. I suppose this special-case handling should be used always when `dl.einsum` is used, right? Therefore, even more so, I would suggest defining an identity element for the distributive laws.

Does this make sense or am I misunderstanding something?

Reply:
I've been trying to come up with a simple unit test for the inconsistent results that were identified in the original issue. However, the issue derives from the lack of a guaranteed order when creating Python sets. There is an assumption about the ordering of the indices within the `remove_messages` function implementation with regard to the ordering of the indices. The use of the `set` function prior to calling `remove_messages` produces inconsistent behavior. So, the incorrect results are only produced occasionally.

There are 2 possible fixes:

1) Use NumPy's set functions in place of the standard Python `set` function so that the indices are always ordered prior to being used as an argument to `remove_message`.

2) Rely on `numpy.einsum` to compute the product of the messages with the message from the current neighbor being excluded. I originally wrote the `remove_message` function in order to avoid repeated calculations of the product of the messages. But the complexity of the code for performing the calculation to divide out the neighbor's message may not be worth any small efficiency gains from avoiding the repeated calculations. 

My commit included both replacing the standard `set` function calls with NumPy set implementations and removing the `remove_message` function to simply the code.

Reply:
Added an additional commit in this branch to support an identity element for SumProduct distributive law

Reply:
Ok! If it's not easy to add a unit test, then that can be left out. :+1: 

Your most recent commit added new files `computation.py` and `sum_product.py` that are in the root of the repository, not inside `junctiontree` package. I suppose this was a mistake, right? Being outside the package, they aren't used anywhere. Perhaps you meant to replace `junctiontree/computation.py` and `junctiontree/sum_product.py`?